File: src/main/java/org/cubexmc/metro/command/newcmd/PortalCommand.java
package org.cubexmc.metro.command.newcmd;

import org.incendo.cloud.annotations.Argument;
import org.incendo.cloud.annotations.CommandDescription;
import org.incendo.cloud.annotations.Command;
import org.incendo.cloud.annotations.Permission;
import org.bukkit.Bukkit;
import org.bukkit.Location;
import org.bukkit.OfflinePlayer;
import org.bukkit.command.CommandSender;
import org.bukkit.entity.Player;
import org.cubexmc.metro.Metro;
import org.cubexmc.metro.manager.LanguageManager;
import org.cubexmc.metro.manager.PortalManager;
import org.cubexmc.metro.model.Portal;
import org.cubexmc.metro.service.CommandDisplayService;
import org.cubexmc.metro.service.PortalCommandService;
import org.cubexmc.metro.update.DataFileUpdater;
import org.cubexmc.metro.util.OwnershipUtil;

import java.util.List;

/**
 * 矿车传送门管理命令。
 */
public class PortalCommand {

    private static final List<String> HELP_KEYS = List.of(
            "portal.help_create",
            "portal.help_setdest",
            "portal.help_link",
            "portal.help_delete",
            "portal.help_list",
            "portal.help_trust",
            "portal.help_untrust",
            "portal.help_owner",
            "portal.help_reload"
    );

    private final Metro plugin;
    private final PortalManager portalManager;
    private final CommandDisplayService displayService;
    private final PortalCommandService portalService;
    private final CommandGuard guard;

    public PortalCommand(Metro plugin) {
        this.plugin = plugin;
        this.portalManager = plugin.getPortalManager();
        this.displayService = new CommandDisplayService();
        this.portalService = new PortalCommandService(portalManager);
        this.guard = new CommandGuard(plugin, plugin.getLineManager(), plugin.getStopManager());
    }

    @Command("m|metro portal")
    @CommandDescription("显示传送门管理帮助")
    public void help(CommandSender sender) {
        showHelp(sender);
    }

    @Command("m|metro portal help")
    @CommandDescription("显示传送门管理帮助")
    public void helpPage(CommandSender sender) {
        showHelp(sender);
    }

    private void showHelp(CommandSender sender) {
        org.cubexmc.metro.manager.LanguageManager lang = plugin.getLanguageManager();
        CommandDisplayService.HelpSection help = displayService.helpSection(key -> lang.getMessage(key),
                "portal.help_header", HELP_KEYS);
        sender.sendMessage(help.header());
        for (String helpLine : help.lines()) {
            sender.sendMessage(helpLine);
        }
    }

    @Command("m|metro portal create <portalId>")
    @CommandDescription("Create a portal entrance at the current position")
    public void createPortal(Player sender, @Argument("portalId") String id) {
        if (!OwnershipUtil.canCreatePortal(sender)) {
            sender.sendMessage(plugin.getLanguageManager().getMessage("portal.permission_create"));
            return;
        }

        PortalCommandService.PortalWriteResult result =
                portalService.createPortal(id, sender.getLocation(), sender.getTargetBlockExact(5),
                        sender.getUniqueId());
        switch (result.status()) {
            case SUCCESS -> {
                Location loc = result.location();
                sender.sendMessage(plugin.getLanguageManager().getMessage("portal.create_success",
                        LanguageManager.put(LanguageManager.put(LanguageManager.put(LanguageManager.put(LanguageManager.put(
                                LanguageManager.args(), "portal_id", id), "x", String.valueOf(loc.getBlockX())),
                                "y", String.valueOf(loc.getBlockY())), "z", String.valueOf(loc.getBlockZ())),
                                "world", loc.getWorld().getName())));
                sender.sendMessage(plugin.getLanguageManager().getMessage("portal.create_setdest_hint",
                        LanguageManager.put(LanguageManager.args(), "portal_id", id)));
            }
            case INVALID_ID -> sender.sendMessage(plugin.getLanguageManager().getMessage("portal.id_invalid",
                    LanguageManager.put(LanguageManager.args(), "portal_id", id)));
            case EXISTS -> sender.sendMessage(plugin.getLanguageManager().getMessage("portal.create_exists",
                    LanguageManager.put(LanguageManager.args(), "portal_id", id)));
            default -> sender.sendMessage(plugin.getLanguageManager().getMessage("portal.create_fail",
                    LanguageManager.put(LanguageManager.args(), "portal_id", id)));
        }
    }

    @Command("m|metro portal setdest <portalId>")
    @CommandDescription("Set the current position as the portal destination")
    public void setDestination(Player sender, @Argument(value = "portalId", suggestions = "portalIds") String id) {
        Portal portal = guard.requireManageablePortal(sender, id);
        if (portal == null) {
            return;
        }

        PortalCommandService.PortalWriteResult result = portalService.setDestination(id, sender.getLocation());
        if (result.status() == PortalCommandService.WriteStatus.NOT_FOUND) {
            sender.sendMessage(plugin.getLanguageManager().getMessage("portal.not_found",
                    LanguageManager.put(LanguageManager.args(), "portal_id", id)));
            return;
        }
        if (result.status() != PortalCommandService.WriteStatus.SUCCESS) {
            sender.sendMessage(plugin.getLanguageManager().getMessage("portal.setdest_fail",
                    LanguageManager.put(LanguageManager.args(), "portal_id", id)));
            return;
        }

        Location loc = result.location();
        sender.sendMessage(plugin.getLanguageManager().getMessage("portal.setdest_success",
                LanguageManager.put(LanguageManager.put(LanguageManager.put(LanguageManager.put(LanguageManager.put(LanguageManager.put(
                        LanguageManager.args(), "portal_id", id), "x", String.format("%.1f", loc.getX())),
                        "y", String.format("%.1f", loc.getY())), "z", String.format("%.1f", loc.getZ())),
                        "yaw", String.format("%.1f", loc.getYaw())), "world", loc.getWorld().getName())));
    }

    @Command("m|metro portal link <id1> <id2>")
    @CommandDescription("双向配对两个传送门")
    public void linkPortals(Player sender,
                            @Argument(value = "id1", suggestions = "portalIds") String id1,
                            @Argument(value = "id2", suggestions = "portalIds") String id2) {
        Portal portal1 = guard.requireManageablePortal(sender, id1);
        if (portal1 == null) {
            return;
        }
        Portal portal2 = guard.requireManageablePortal(sender, id2);
        if (portal2 == null) {
            return;
        }

        if (portalService.linkPortals(id1, id2) != PortalCommandService.WriteStatus.SUCCESS) {
            sender.sendMessage(plugin.getLanguageManager().getMessage("portal.link_fail"));
            return;
        }
        sender.sendMessage(plugin.getLanguageManager().getMessage("portal.link_success",
                LanguageManager.put(LanguageManager.put(LanguageManager.args(), "portal_id_1", id1), "portal_id_2", id2)));
    }

    @Command("m|metro portal delete <portalId> [confirm]")
    @CommandDescription("Delete a portal")
    public void deletePortal(Player sender,
                             @Argument(value = "portalId", suggestions = "portalIds") String id,
                             @Argument("confirm") String confirm) {
        Portal portal = guard.requireManageablePortal(sender, id);
        if (portal == null) {
            return;
        }
        if (!guard.requireConfirmation(sender, confirm, "/m portal delete " + id + " confirm")) {
            return;
        }

        if (portalService.deletePortal(id) != PortalCommandService.WriteStatus.SUCCESS) {
            sender.sendMessage(plugin.getLanguageManager().getMessage("portal.not_found",
                    LanguageManager.put(LanguageManager.args(), "portal_id", id)));
            return;
        }
        sender.sendMessage(plugin.getLanguageManager().getMessage("portal.delete_success",
                LanguageManager.put(LanguageManager.args(), "portal_id", id)));
    }

    @Command("m|metro portal list [page]")
    @CommandDescription("列出所有传送门")
    public void listPortals(CommandSender sender, @Argument(value = "page", suggestions = "pageNumbers") Integer page) {
        List<Portal> allPortals = portalService.listPortals();
        if (allPortals.isEmpty()) {
            sender.sendMessage(plugin.getLanguageManager().getMessage("portal.list_empty"));
            return;
        }

        CommandDisplayService.Page<Portal> portalPage = displayService.paginate(allPortals, page);
        sender.sendMessage(displayService.pageHeader(
                plugin.getLanguageManager().getMessage("portal.list_header",
                        LanguageManager.put(LanguageManager.args(), "count", String.valueOf(allPortals.size()))),
                portalPage));
        for (Portal p : portalPage.items()) {
            String linked = p.getLinkedPortalId() != null
                    ? plugin.getLanguageManager().getMessage("portal.list_linked",
                            LanguageManager.put(LanguageManager.args(), "linked_portal_id", p.getLinkedPortalId()))
                    : "";
            sender.sendMessage(plugin.getLanguageManager().getMessage("portal.list_item",
                    LanguageManager.put(LanguageManager.put(LanguageManager.put(LanguageManager.put(LanguageManager.put(
                            LanguageManager.put(LanguageManager.put(LanguageManager.put(LanguageManager.args(),
                                    "portal_id", p.getId()), "world", p.getWorldName()),
                                    "x", String.valueOf(p.getX())), "y", String.valueOf(p.getY())),
                            "z", String.valueOf(p.getZ())), "dest_world", p.getDestWorldName()),
                            "dest", String.format("%.0f,%.0f,%.0f", p.getDestX(), p.getDestY(), p.getDestZ())),
                            "linked", linked)));
        }
    }

    @Command("m|metro portal trust <portalId> <playerName>")
    @CommandDescription("Grant portal admin permissions")
    public void trust(Player player,
                      @Argument(value = "portalId", suggestions = "portalIds") String id,
                      @Argument(value = "playerName", suggestions = "playerNames") String playerName) {
        Portal portal = guard.requireManageablePortal(player, id);
        if (portal == null) {
            return;
        }
        OfflinePlayer target = Bukkit.getOfflinePlayer(playerName);
        if (!target.hasPlayedBefore() && !target.isOnline()) {
            player.sendMessage(plugin.getLanguageManager().getMessage("command.player_not_found",
                    LanguageManager.put(LanguageManager.args(), "player", playerName)));
            return;
        }
        PortalCommandService.WriteStatus status = portalService.addAdmin(portal, target.getUniqueId());
        if (status == PortalCommandService.WriteStatus.EXISTS) {
            player.sendMessage(plugin.getLanguageManager().getMessage("portal.trust_exists",
                    LanguageManager.put(LanguageManager.args(), "player", playerName)));
        } else if (status == PortalCommandService.WriteStatus.SUCCESS) {
            player.sendMessage(plugin.getLanguageManager().getMessage("portal.trust_success",
                    LanguageManager.put(LanguageManager.put(LanguageManager.args(), "portal_id", id), "player", playerName)));
        } else {
            player.sendMessage(plugin.getLanguageManager().getMessage("portal.trust_fail",
                    LanguageManager.put(LanguageManager.put(LanguageManager.args(), "portal_id", id), "player", playerName)));
        }
    }

    @Command("m|metro portal untrust <portalId> <playerName>")
    @CommandDescription("Remove portal admin permissions")
    public void untrust(Player player,
                        @Argument(value = "portalId", suggestions = "portalIds") String id,
                        @Argument(value = "playerName", suggestions = "playerNames") String playerName) {
        Portal portal = guard.requireManageablePortal(player, id);
        if (portal == null) {
            return;
        }
        OfflinePlayer target = Bukkit.getOfflinePlayer(playerName);
        if (!target.hasPlayedBefore() && !target.isOnline()) {
            player.sendMessage(plugin.getLanguageManager().getMessage("command.player_not_found",
                    LanguageManager.put(LanguageManager.args(), "player", playerName)));
            return;
        }
        PortalCommandService.WriteStatus status = portalService.removeAdmin(portal, target.getUniqueId());
        if (status == PortalCommandService.WriteStatus.SUCCESS) {
            player.sendMessage(plugin.getLanguageManager().getMessage("portal.untrust_success",
                    LanguageManager.put(LanguageManager.put(LanguageManager.args(), "portal_id", id), "player", playerName)));
        } else {
            player.sendMessage(plugin.getLanguageManager().getMessage("portal.untrust_fail",
                    LanguageManager.put(LanguageManager.put(LanguageManager.args(), "portal_id", id), "player", playerName)));
        }
    }

    @Command("m|metro portal owner <portalId> <playerName>")
    @CommandDescription("转移传送门所有权")
    public void owner(Player player,
                      @Argument(value = "portalId", suggestions = "portalIds") String id,
                      @Argument(value = "playerName", suggestions = "playerNames") String playerName) {
        Portal portal = guard.requireManageablePortal(player, id);
        if (portal == null || !guard.requirePortalOwner(player, portal)) {
            return;
        }
        OfflinePlayer target = Bukkit.getOfflinePlayer(playerName);
        if (!target.hasPlayedBefore() && !target.isOnline()) {
            player.sendMessage(plugin.getLanguageManager().getMessage("command.player_not_found",
                    LanguageManager.put(LanguageManager.args(), "player", playerName)));
            return;
        }
        PortalCommandService.WriteStatus status = portalService.setOwner(portal, target.getUniqueId());
        if (status == PortalCommandService.WriteStatus.SUCCESS) {
            player.sendMessage(plugin.getLanguageManager().getMessage("portal.owner_success",
                    LanguageManager.put(LanguageManager.put(LanguageManager.args(), "portal_id", id), "owner", playerName)));
        } else {
            player.sendMessage(plugin.getLanguageManager().getMessage("portal.owner_fail",
                    LanguageManager.put(LanguageManager.args(), "portal_id", id)));
        }
    }

    @Command("m|metro portal reload")
    @CommandDescription("重新加载传送门配置")
    @Permission("railway.admin")
    public void reloadPortals(CommandSender sender) {
        PortalCommandService.ReloadResult result =
                portalService.reloadPortals(() -> DataFileUpdater.migratePortals(plugin));
        sender.sendMessage(plugin.getLanguageManager().getMessage("portal.reload_success",
                LanguageManager.put(LanguageManager.args(), "count", String.valueOf(result.portalCount()))));
    }
}

File: src/main/java/org/cubexmc/metro/command/newcmd/PortalCommand.kt
package org.cubexmc.metro.command.newcmd

import org.bukkit.Bukkit
import org.bukkit.Location
import org.bukkit.command.CommandSender
import org.bukkit.entity.Player
import org.cubexmc.metro.Metro
import org.cubexmc.metro.manager.LanguageManager
import org.cubexmc.metro.model.Portal
import org.cubexmc.metro.service.CommandDisplayService
import org.cubexmc.metro.service.PortalCommandService
import org.cubexmc.metro.update.DataFileUpdater
import org.cubexmc.metro.util.OwnershipUtil
import org.incendo.cloud.annotations.Argument
import org.incendo.cloud.annotations.Command
import org.incendo.cloud.annotations.CommandDescription
import org.incendo.cloud.annotations.Permission
import java.util.function.Function

/**
 * 矿车传送门管理命令。
 */
class PortalCommand(private val plugin: Metro) {

    private val displayService = CommandDisplayService()
    private val portalService = PortalCommandService(plugin.portalManager)
    private val guard = CommandGuard(plugin, plugin.lineManager, plugin.stopManager)

    @Command("rw|railway|rail portal")
    @CommandDescription("显示传送门管理帮助")
    fun help(sender: CommandSender) {
        showHelp(sender)
    }

    @Command("rw|railway|rail portal help")
    @CommandDescription("显示传送门管理帮助")
    fun helpPage(sender: CommandSender) {
        showHelp(sender)
    }

    private fun showHelp(sender: CommandSender) {
        val lang = plugin.languageManager
        val help =
            displayService.helpSection(
                Function { key -> lang.getMessage(key) },
                "portal.help_header",
                HELP_KEYS,
            )
        sender.sendMessage(help.header())
        for (helpLine in help.lines()) {
            sender.sendMessage(helpLine)
        }
    }

    @Command("rw|railway|rail portal create <portalId>")
    @CommandDescription("Create a portal entrance at the current position")
    fun createPortal(sender: Player, @Argument("portalId") id: String) {
        if (!OwnershipUtil.canCreatePortal(sender)) {
            sender.sendMessage(plugin.languageManager.getMessage("portal.permission_create"))
            return
        }

        val result =
            portalService.createPortal(
                id,
                sender.location,
                sender.getTargetBlockExact(TARGET_BLOCK_RANGE),
                sender.uniqueId,
            )
        when (result.status) {
            PortalCommandService.WriteStatus.SUCCESS -> {
                val loc = result.location ?: return
                val args = LanguageManager.args()
                LanguageManager.put(args, "portal_id", id)
                LanguageManager.put(args, "x", loc.blockX.toString())
                LanguageManager.put(args, "y", loc.blockY.toString())
                LanguageManager.put(args, "z", loc.blockZ.toString())
                LanguageManager.put(args, "world", worldName(loc))
                sender.sendMessage(plugin.languageManager.getMessage("portal.create_success", args))
                sender.sendMessage(
                    plugin.languageManager.getMessage(
                        "portal.create_setdest_hint",
                        LanguageManager.put(LanguageManager.args(), "portal_id", id),
                    ),
                )
            }

            PortalCommandService.WriteStatus.INVALID_ID ->
                sender.sendMessage(
                    plugin.languageManager.getMessage(
                        "portal.id_invalid",
                        LanguageManager.put(LanguageManager.args(), "portal_id", id),
                    ),
                )

            PortalCommandService.WriteStatus.EXISTS ->
                sender.sendMessage(
                    plugin.languageManager.getMessage(
                        "portal.create_exists",
                        LanguageManager.put(LanguageManager.args(), "portal_id", id),
                    ),
                )

            else ->
                sender.sendMessage(
                    plugin.languageManager.getMessage(
                        "portal.create_fail",
                        LanguageManager.put(LanguageManager.args(), "portal_id", id),
                    ),
                )
        }
    }

    @Command("rw|railway|rail portal setdest <portalId>")
    @CommandDescription("Set the current position as the portal destination")
    fun setDestination(
        sender: Player,
        @Argument(value = "portalId", suggestions = "portalIds") id: String,
    ) {
        guard.requireManageablePortal(sender, id) ?: return

        val result = portalService.setDestination(id, sender.location)
        if (result.status == PortalCommandService.WriteStatus.NOT_FOUND) {
            sender.sendMessage(
                plugin.languageManager.getMessage(
                    "portal.not_found",
                    LanguageManager.put(LanguageManager.args(), "portal_id", id),
                ),
            )
            return
        }
        if (result.status != PortalCommandService.WriteStatus.SUCCESS) {
            sender.sendMessage(
                plugin.languageManager.getMessage(
                    "portal.setdest_fail",
                    LanguageManager.put(LanguageManager.args(), "portal_id", id),
                ),
            )
            return
        }

        val loc = result.location ?: return
        val args = LanguageManager.args()
        LanguageManager.put(args, "portal_id", id)
        LanguageManager.put(args, "x", String.format("%.1f", loc.x))
        LanguageManager.put(args, "y", String.format("%.1f", loc.y))
        LanguageManager.put(args, "z", String.format("%.1f", loc.z))
        LanguageManager.put(args, "yaw", String.format("%.1f", loc.yaw))
        LanguageManager.put(args, "world", worldName(loc))
        sender.sendMessage(plugin.languageManager.getMessage("portal.setdest_success", args))
    }

    @Command("rw|railway|rail portal link <id1> <id2>")
    @CommandDescription("双向配对两个传送门")
    fun linkPortals(
        sender: Player,
        @Argument(value = "id1", suggestions = "portalIds") id1: String,
        @Argument(value = "id2", suggestions = "portalIds") id2: String,
    ) {
        guard.requireManageablePortal(sender, id1) ?: return
        guard.requireManageablePortal(sender, id2) ?: return
        if (portalService.linkPortals(id1, id2) != PortalCommandService.WriteStatus.SUCCESS) {
            sender.sendMessage(plugin.languageManager.getMessage("portal.link_fail"))
            return
        }
        val args = LanguageManager.args()
        LanguageManager.put(args, "portal_id_1", id1)
        LanguageManager.put(args, "portal_id_2", id2)
        sender.sendMessage(plugin.languageManager.getMessage("portal.link_success", args))
    }

    @Command("rw|railway|rail portal delete <portalId> [confirm]")
    @CommandDescription("Delete a portal")
    fun deletePortal(
        sender: Player,
        @Argument(value = "portalId", suggestions = "portalIds") id: String,
        @Argument("confirm") confirm: String?,
    ) {
        guard.requireManageablePortal(sender, id) ?: return
        if (!guard.requireConfirmation(sender, confirm, "/rail portal delete $id confirm")) {
            return
        }

        if (portalService.deletePortal(id) != PortalCommandService.WriteStatus.SUCCESS) {
            sender.sendMessage(
                plugin.languageManager.getMessage(
                    "portal.not_found",
                    LanguageManager.put(LanguageManager.args(), "portal_id", id),
                ),
            )
            return
        }
        sender.sendMessage(
            plugin.languageManager.getMessage(
                "portal.delete_success",
                LanguageManager.put(LanguageManager.args(), "portal_id", id),
            ),
        )
    }

    @Command("rw|railway|rail portal list [page]")
    @CommandDescription("列出所有传送门")
    fun listPortals(
        sender: CommandSender,
        @Argument(value = "page", suggestions = "pageNumbers") page: Int?,
    ) {
        val allPortals = portalService.listPortals()
        if (allPortals.isEmpty()) {
            sender.sendMessage(plugin.languageManager.getMessage("portal.list_empty"))
            return
        }

        val portalPage = displayService.paginate(allPortals, page)
        sender.sendMessage(
            displayService.pageHeader(
                plugin.languageManager.getMessage(
                    "portal.list_header",
                    LanguageManager.put(LanguageManager.args(), "count", allPortals.size.toString()),
                ),
                portalPage,
            ),
        )
        for (portal in portalPage.items()) {
            sender.sendMessage(plugin.languageManager.getMessage("portal.list_item", listItemArgs(portal)))
        }
    }

    private fun listItemArgs(portal: Portal): MutableMap<String, Any?> {
        val args = LanguageManager.args()
        LanguageManager.put(args, "portal_id", portal.id)
        LanguageManager.put(args, "world", portal.worldName)
        LanguageManager.put(args, "x", portal.x.toString())
        LanguageManager.put(args, "y", portal.y.toString())
        LanguageManager.put(args, "z", portal.z.toString())
        LanguageManager.put(args, "dest_world", portal.destWorldName)
        LanguageManager.put(
            args,
            "dest",
            String.format("%.0f,%.0f,%.0f", portal.destX, portal.destY, portal.destZ),
        )
        val linked = portal.linkedPortalId?.let { linkedPortalId ->
            plugin.languageManager.getMessage(
                "portal.list_linked",
                LanguageManager.put(LanguageManager.args(), "linked_portal_id", linkedPortalId),
            )
        }.orEmpty()
        LanguageManager.put(args, "linked", linked)
        return args
    }

    @Command("rw|railway|rail portal trust <portalId> <playerName>")
    @CommandDescription("Grant portal admin permissions")
    fun trust(
        player: Player,
        @Argument(value = "portalId", suggestions = "portalIds") id: String,
        @Argument(value = "playerName", suggestions = "playerNames") playerName: String,
    ) {
        val portal = guard.requireManageablePortal(player, id) ?: return
        val target = Bukkit.getOfflinePlayer(playerName)
        if (!target.hasPlayedBefore() && !target.isOnline) {
            player.sendMessage(playerNotFoundMessage(playerName))
            return
        }
        when (portalService.addAdmin(portal, target.uniqueId)) {
            PortalCommandService.WriteStatus.EXISTS ->
                player.sendMessage(
                    plugin.languageManager.getMessage(
                        "portal.trust_exists",
                        LanguageManager.put(LanguageManager.args(), "player", playerName),
                    ),
                )

            PortalCommandService.WriteStatus.SUCCESS ->
                player.sendMessage(
                    plugin.languageManager.getMessage("portal.trust_success", portalPlayerArgs(id, playerName)),
                )

            else ->
                player.sendMessage(
                    plugin.languageManager.getMessage("portal.trust_fail", portalPlayerArgs(id, playerName)),
                )
        }
    }

    @Command("rw|railway|rail portal untrust <portalId> <playerName>")
    @CommandDescription("Remove portal admin permissions")
    fun untrust(
        player: Player,
        @Argument(value = "portalId", suggestions = "portalIds") id: String,
        @Argument(value = "playerName", suggestions = "playerNames") playerName: String,
    ) {
        val portal = guard.requireManageablePortal(player, id) ?: return
        val target = Bukkit.getOfflinePlayer(playerName)
        if (!target.hasPlayedBefore() && !target.isOnline) {
            player.sendMessage(playerNotFoundMessage(playerName))
            return
        }
        if (portalService.removeAdmin(portal, target.uniqueId) == PortalCommandService.WriteStatus.SUCCESS) {
            player.sendMessage(
                plugin.languageManager.getMessage("portal.untrust_success", portalPlayerArgs(id, playerName)),
            )
        } else {
            player.sendMessage(
                plugin.languageManager.getMessage("portal.untrust_fail", portalPlayerArgs(id, playerName)),
            )
        }
    }

    @Command("rw|railway|rail portal owner <portalId> <playerName>")
    @CommandDescription("转移传送门所有权")
    fun owner(
        player: Player,
        @Argument(value = "portalId", suggestions = "portalIds") id: String,
        @Argument(value = "playerName", suggestions = "playerNames") playerName: String,
    ) {
        val portal = guard.requireManageablePortal(player, id)
        if (portal == null || !guard.requirePortalOwner(player, portal)) {
            return
        }
        val target = Bukkit.getOfflinePlayer(playerName)
        if (!target.hasPlayedBefore() && !target.isOnline) {
            player.sendMessage(playerNotFoundMessage(playerName))
            return
        }
        if (portalService.setOwner(portal, target.uniqueId) == PortalCommandService.WriteStatus.SUCCESS) {
            val args = LanguageManager.args()
            LanguageManager.put(args, "portal_id", id)
            LanguageManager.put(args, "owner", playerName)
            player.sendMessage(plugin.languageManager.getMessage("portal.owner_success", args))
        } else {
            player.sendMessage(
                plugin.languageManager.getMessage(
                    "portal.owner_fail",
                    LanguageManager.put(LanguageManager.args(), "portal_id", id),
                ),
            )
        }
    }

    @Command("rw|railway|rail portal reload")
    @CommandDescription("重新加载传送门配置")
    @Permission("railway.admin")
    fun reloadPortals(sender: CommandSender) {
        val result = portalService.reloadPortals { DataFileUpdater.migratePortals(plugin) }
        sender.sendMessage(
            plugin.languageManager.getMessage(
                "portal.reload_success",
                LanguageManager.put(LanguageManager.args(), "count", result.portalCount.toString()),
            ),
        )
    }

    private fun playerNotFoundMessage(playerName: String): String =
        plugin.languageManager.getMessage(
            "command.player_not_found",
            LanguageManager.put(LanguageManager.args(), "player", playerName),
        )

    private fun portalPlayerArgs(portalId: String, playerName: String): MutableMap<String, Any?> {
        val args = LanguageManager.args()
        LanguageManager.put(args, "portal_id", portalId)
        LanguageManager.put(args, "player", playerName)
        return args
    }

    private fun worldName(location: Location): String = location.world?.name ?: ""

    private companion object {
        const val TARGET_BLOCK_RANGE = 5

        val HELP_KEYS =
            listOf(
                "portal.help_create",
                "portal.help_setdest",
                "portal.help_link",
                "portal.help_delete",
                "portal.help_list",
                "portal.help_trust",
                "portal.help_untrust",
                "portal.help_owner",
                "portal.help_reload",
            )
    }
}
